Move psutil to test requirements and do optimistic load

B-Step62 opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/B-Step62/mlflow/pull/10342?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10342/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10342
```

</p>
</details>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?
`psutil` doesn't provide Arm64 Linux binary, requiring building from CPython source which is a headache. This change makes it optional and show user-friendly warning when user tries to run system metrics related features without installing it.

Error message will look like this;
```
  File "mlflow/utils/process.py", line 224, in <module>
    _check_and_import_psutil()
  File "mlflow/utils/process.py", line 172, in _check_and_import_psutil
    raise ImportError(
ImportError: No module named 'psutil'

HINT: The `psutil` package is required to log system metrics in MLflow. Run 
`pip install psutil --upgrade` to install the package. If you are on an Arm64-based
 Linux system, you may need to have GCC installed to build the wheel from source.
```

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [x] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

B-Step62 committed:
Move psutil to test requirements and do optimistic load for system metrics support.

psutil doesn't provide Arm64 Linux binary, requiring building from CPython source which is a headache. This change make it optional and show user-friendly warning when user tries to run system metrics related features without installing it.

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 commented on conftest.py:
```diff
@@ -104,17 +106,16 @@ def pytest_runtest_setup(item):
-            import psutil
-        except ImportError:
+        # SKip showing stats at all if psutil is not installed
+        if importlib.util.find_spec("psutil") is None:
```
suggested by ruff

B-Step62 committed:
Add psutil to gateway test dependencies

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

harupy commented:
We should also think about how we can prevent using `psutil` directly. For example, if I make the following change, would CI block my PR or not? 

```diff
+ import psutil
+
+ print(psutil.__version__)
```

B-Step62 commented:
> We should also think about how we can prevent using `psutil` directly. For example, if I make the following change, would CI block my PR or not?
> 
> ```diff
> + import psutil
> +
> + print(psutil.__version__)
> ```

Makes sense, probably adding custom pylint checker would be the easiest path. Let me give it a try.

B-Step62 committed:
Add pylint checker for blocking direct import of psutil

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

harupy commented:
Have we tried using `LazyLoader`?

```python
from mlflow.utils.lazy_load import LazyLoader

psutil = LazyLoader("psutil", globals(), "psutil")

print(psutil.cpu_count())
```

B-Step62 commented:
> Have we tried using `LazyLoader`?

Ah... didn't know that class. This solves the issue then.

B-Step62 commented:
> Have we tried using `LazyLoader`?
> 
> ```python
> from mlflow.utils.lazy_load import LazyLoader
> 
> psutil = LazyLoader("psutil", globals(), "psutil")
> 
> print(psutil.cpu_count())
> ```

Ah.... actually `psutil` usage is pretty much inside `system_metric` module and `gateway` outside tests.... so it sounds solve the issue.

B-Step62 committed:
Remove wrapper. psutil was actually only used by system metrics
module and gateway, both are already lazy loaded. Hence, just
added some check and suggestion.

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 commented:
> > Have we tried using `LazyLoader`?
> > ```python
> > from mlflow.utils.lazy_load import LazyLoader
> > 
> > psutil = LazyLoader("psutil", globals(), "psutil")
> > 
> > print(psutil.cpu_count())
> > ```
> 
> Ah.... actually `psutil` usage is pretty much inside `system_metric` module and `gateway` outside tests.... so it sounds solve the issue at all.

Just found that `SystemMonitorMetrics` is lazy loaded in current implementation anyway ([code](https://github.com/mlflow/mlflow/blob/e90a2c8ee0160006a83930f68899d8771e381877/mlflow/tracking/fluent.py#L384C1-L385C1)), so probably we don't need to use LazyLoader for it but just keep it and add some helpful message there about psutil.

harupy commented on requirements/test-requirements.txt:
```diff
@@ -42,3 +42,5 @@ databricks-sdk
 # TODO: Unpin once we support openai 1.0
 openai<1.0
+# Required for showing pytest stats
+psutil<6
```
```suggestion
psutil
```

B-Step62 replied:
Got it, will remove

harupy commented on mlflow/tracking/fluent.py:
```diff
@@ -380,6 +380,17 @@ def start_run(
             log_system_metrics = MLFLOW_ENABLE_SYSTEM_METRICS_LOGGING.get()
         if log_system_metrics:
+            # Ensure psutil is installed. It was moved to an optional dependency, as it doesn't
+            # have binary for Arm64 Linux and requires build from CPython which was a headache.
```
Any issue we can link?

harupy commented on mlflow/tracking/fluent.py:
```diff
@@ -380,6 +380,17 @@ def start_run(
+            if importlib.util.find_spec("psutil") is None:
+                raise MlflowException(
+                    "Failed to start system metrics monitoring as package `psutil` is not "
+                    "installed. Run `pip install psutil --upgrade` to resolve the issue, "
```
```suggestion
                    "installed. Run `pip install psutil` to resolve the issue, "
```

B-Step62 replied:
ah we don't need, just a random sugestion from copilot

harupy commented on mlflow/tracking/fluent.py:
```diff
@@ -380,6 +380,17 @@ def start_run(
+                    "installed. Run `pip install psutil --upgrade` to resolve the issue, "
+                    "otherwise you can disable system metrics logging by passing "
+                    "`log_system_metrics=False` to start_run() or setting environment "
+                    "variable MLFLOW_ENABLE_SYSTEM_METRICS_LOGGING to False."
```
```suggestion
                    "environment variable MLFLOW_ENABLE_SYSTEM_METRICS_LOGGING to False."
```

B-Step62 replied:
oh do you mean changing line break?

harupy replied:
Never mind. We already have `environment`. I missed it.

B-Step62 committed:
comments

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

harupy reviewed (approved):
LGTM

harupy commented on mlflow/tracking/fluent.py:
```diff
@@ -380,6 +380,18 @@ def start_run(
+                    "installed. Run `pip install psutil` to resolve the issue, "
+                    "otherwise you can disable system metrics logging by passing "
+                    "`log_system_metrics=False` to start_run() or setting environment "
+                    "variable MLFLOW_ENABLE_SYSTEM_METRICS_LOGGING to False."
```
```suggestion
                    f"variable {MLFLOW_ENABLE_SYSTEM_METRICS_LOGGING} to False."
```

harupy commented on mlflow/tracking/fluent.py:
```diff
@@ -380,6 +380,18 @@ def start_run(
             log_system_metrics = MLFLOW_ENABLE_SYSTEM_METRICS_LOGGING.get()
         if log_system_metrics:
+            # Ensure psutil is installed. It was moved to an optional dependency, as it doesn't
+            # have binary for Arm64 Linux and requires build from CPython which was a headache.
```
```suggestion
            # have binary for Arm64 Linux and requires build from CPython which is a headache.
```

`was a headache` sounds `it is not now`.

harupy commented on mlflow/tracking/fluent.py:
```diff
@@ -380,6 +380,18 @@ def start_run(
+                    "Failed to start system metrics monitoring as package `psutil` is not "
+                    "installed. Run `pip install psutil` to resolve the issue, "
+                    "otherwise you can disable system metrics logging by passing "
+                    "`log_system_metrics=False` to start_run() or setting environment "
```
```suggestion
                    "`log_system_metrics=False` to `start_run()` or setting environment "
```

B-Step62 committed:
Update mlflow/tracking/fluent.py

Co-authored-by: Harutaka Kawamura <hkawamura0130@gmail.com>
Signed-off-by: Yuki Watanabe <31463517+B-Step62@users.noreply.github.com>

B-Step62 committed:
Update mlflow/tracking/fluent.py

Co-authored-by: Harutaka Kawamura <hkawamura0130@gmail.com>
Signed-off-by: Yuki Watanabe <31463517+B-Step62@users.noreply.github.com>

B-Step62 committed:
Update mlflow/tracking/fluent.py

Co-authored-by: Harutaka Kawamura <hkawamura0130@gmail.com>
Signed-off-by: Yuki Watanabe <31463517+B-Step62@users.noreply.github.com>
